// @popperjs/core/lib/enums.js
// Placements
export const direction = {
  top: 'top',
  bottom: 'bottom',
  right: 'right',
  left: 'left',
  auto: 'auto',
};

export const basePlacements = [direction.top, direction.bottom, direction.right, direction.left];

export const start = 'start';
export const end = 'end';

export const variationPlacements = new Set(
  basePlacements.flatMap((placement) => [
    `${placement}-${start}`,
    `${placement}-${end}`,
  ])
);

export const placements = new Set([
  ...basePlacements,
  ...basePlacements.flatMap((placement) => [
    `${placement}-${start}`,
    `${placement}-${end}`,
  ]),
  direction.auto,
]);

// Modifier Phases
export const beforeRead = 'beforeRead';
export const read = 'read';
export const afterRead = 'afterRead';

export const beforeMain = 'beforeMain';
export const main = 'main';
export const afterMain = 'afterMain';

export const beforeWrite = 'beforeWrite';
export const write = 'write';
export const afterWrite = 'afterWrite';

export const modifierPhases = new Set([
  beforeRead,
  read,
  afterRead,
  beforeMain,
  main,
  afterMain,
  beforeWrite,
  write,
  afterWrite,
]);

// Other exports
export const clippingParents = 'clippingParents';
export const viewport = 'viewport';
export const popper = 'popper';
export const reference = 'reference';
